cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringServerHandler.java

```diff
@@ -0,0 +1,231 @@
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.QueryParam;
+
+/**
+ * {@link io.cdap.http.HttpHandler} to manage tethering server v3 REST APIs
+ */
+@Path(Constants.Gateway.API_VERSION_3)
+public class TetheringServerHandler extends AbstractHttpHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(TetheringServerHandler.class);
+  private static final Gson GSON = new GsonBuilder().create();
+  private final CConfiguration cConf;
+  private final TetheringStore store;
+  private final MessagingService messagingService;
+  private final MultiThreadMessagingContext messagingContext;
+  private final String topicPrefix;
+
+  @Inject
+  TetheringServerHandler(CConfiguration cConf, TetheringStore store, MessagingService messagingService) {
+    this.cConf = cConf;
+    this.store = store;
+    this.messagingService = messagingService;
+    this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    this.topicPrefix = cConf.get(Constants.Tethering.TOPIC_PREFIX);
+  }
+
+  @Override
+  public void init(HandlerContext context) {
+    super.init(context);
+  }
+
+  /**
+   * Sends control commands to the client.
+   */
+  @GET
+  @Path("/tethering/controlchannels/{peer}")
+  public void connectControlChannel(FullHttpRequest request, HttpResponder responder, @PathParam("peer") String peer,
+                                    @QueryParam("messageId") String messageId)
+    throws IOException, NotImplementedException, PeerNotFoundException, ForbiddenException, BadRequestException {
+    checkTetheringServerEnabled();
+    store.updatePeerTimestamp(peer);
+    TetheringStatus tetheringStatus = store.getPeer(peer).getTetheringStatus();
+    if (tetheringStatus == TetheringStatus.PENDING) {
+      throw new PeerNotFoundException(String.format("Peer %s not found", peer));
+    } else if (tetheringStatus == TetheringStatus.REJECTED) {
+      responder.sendStatus(HttpResponseStatus.FORBIDDEN);
+      throw new ForbiddenException(String.format("Peer %s is not authorized", peer));
+    }
+
+    List<TetheringControlMessage> commands = new ArrayList<>();
+    MessageFetcher fetcher = messagingContext.getMessageFetcher();
+    TopicId topic = new TopicId(NamespaceId.SYSTEM.getNamespace(), topicPrefix + peer);
+    String lastMessageId = messageId;
+    try (CloseableIterator<Message> iterator =
+           fetcher.fetch(topic.getNamespace(), topic.getTopic(), 1, messageId)) {
+      while (iterator.hasNext()) {
```

Comment:
Typically we don't expect a lot of messages, do we? Consider using BodyProducer if the response can be large.

Reply:
Don't expect a lot of messages - should just send messages when the user runs or kills a pipeline.
Changed to use BodyProducer